### PR TITLE
Kareem branch

### PR DESCRIPTION
**Pull request by KareemSalem7**
Finished tree intent functions and completed one comprehensive test for tree intents and getting two lists of intents

### CODE REVIEW

**Comment by MartinCalcaterra** [comment]
Just resolved merge conflicts between Kareem branch and main- Clean architecture changes in main preserved, and combined with business logic updates in Kareem branch.

It would be a good idea to review the changes I made in that commit (a983091) to make sure the clean architecture is all good. The business logic tests are working just fine.

Also, in the commit directly after I resolved an error relating to a redundant and outdated Tree.java file, which has been removed.

File: transcript/src/test/java/com/transcriptanalyzer/transcript/TreeTest.java [hunk 1]
```diff
@@ -1,48 +1,72 @@
...
+        turn1T1.add("Remake");
+        transcript1.add(turn1T1);
+        turn2T1.add("Sorry, could you rephrase that for me?");
+        turn2T1.add("Order");
+        transcript1.add(turn2T1);
+        turn3T1.add("We have 3 different kinds of pizzas available with one being on special per day. The choices are chesse, veggie, and pepperoni.");
+        turn3T1.add("Veggie");
+        transcript1.add(turn3T1);
+        turn4T1.add("We will make you a veggie pizza. Please come pick it up in 20 minutes.");
+        transcript1.add(turn4T1);
```
Comment (CharlieMorocz) [inline transcript/src/test/java/com/transcriptanalyzer/transcript/TreeTest.java:47]: For future testing purposes, it may be worth creating a function which could take in an array of strings and automatically generate these mock transcripts - would clean up the file and make it a lot less tedious to write new tests! Great job on implementing such thorough testing though.

File: transcript/src/main/java/com/transcriptanalyzer/transcript/User_Requests_Intents/Service/TranscriptService.java [hunk 2]
```diff
@@ -112,48 +130,101 @@ private static void flowIterator2(ArrayList<ArrayList<ArrayList<String>>> finalP
...
+
+                turnToMerge.add(String.valueOf(currMessage));
+
+//              Isolate messages of end turns of successful transcripts
+                if (!String.valueOf(currIntent).equals("Null")) {
+                    turnToMerge.add(String.valueOf(currIntent));
+                }
+
+                mergedResult.add(turnToMerge);
+                intentIndex = intentIndex + 2;
```
Comment (CharlieMorocz) [inline transcript/src/main/java/com/transcriptanalyzer/transcript/User_Requests_Intents/Service/TranscriptService.java:206]: Minor change, but we may want to consider extracting this while loop into a helper method to reduce the bulk of this function.

File: transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java [hunk 3]
```diff
@@ -71,17 +78,95 @@ public Node hasChild(String intent){
             }
             return null;
         }
+
+        // return a hashmap mapping the 3 leaf intents with most occurrences to their occurrences in descending order
+        public LinkedHashMap<String, Integer> getBestLeafData() {
```
Comment (CharlieMorocz) [inline transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java:83]: The documentation is fantastic here! It was super easy to follow along with what the function is supposed to be doing at each step - definitely will be helpful if there is any debugging to do in the future!

File: transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java [hunk 4]
```diff
@@ -71,17 +78,95 @@ public Node hasChild(String intent){
...
+            // create a list for the best leaf data of all children
+            LinkedHashMap<String, Integer> leafData = new LinkedHashMap<>();
+            // create a placeholder list for the top three leaf intents
+            LinkedHashMap<String, Integer> bestLeafData = new LinkedHashMap<>();
```
Comment (CharlieMorocz) [inline transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java:87]: We might want to consider dividing up the process of collecting the best leaf data of all children and collecting the top three intents - these two variables have very similar names and method calls so it makes it difficult at times to track which one we're modifying at each step. Helper methods (i.e., one the returns the best leaf data for all children) would be useful here.

File: transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java [hunk 5]
```diff
@@ -71,17 +78,95 @@ public Node hasChild(String intent){
...
+                // determine how many entries to take from the list
+                // take the last index of the entries we will take. no more than 3,
+                // and no more than the number of elements in the list
+                int last_index = java.lang.Math.min(2, leafDataEntries.size() - 1);
+
+                // for each index of the entries we want
+                for (int i = 0; i <= last_index; i++) {
+                    // directly extract the key and value at that index
+                    String key = leafDataEntries.get(i).getKey();
+                    Integer value = leafDataEntries.get(i).getValue();
```
Comment (CharlieMorocz) [inline transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java:129]: This is a really clever way of isolating the top intents while avoiding potential errors - well done!

File: transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java [hunk 6]
```diff
@@ -71,17 +78,95 @@ public Node hasChild(String intent){
...
     }
 
     // get a list of the three nodes with the greatest occurrences sorted in decreasing order
-    public List<String> getTopThree(){
+    public List<String> getBestTreeIntents(){
         // compares each value and get the top 3, return key list attributed to these top 3
         return this.counts.entrySet().stream().sorted(
                 Map.Entry.<String, Integer>comparingByValue().reversed()).limit(3).map(
                         Map.Entry::getKey).collect(Collectors.toList());
     }
```
Comment (CharlieMorocz) [inline transcript/src/main/java/com/transcriptanalyzer/transcript/Tree.java:148]: We should really consider dividing this up somehow as it's rather deeply nested. Good job coming up with this however haha!

File: transcript/src/main/java/com/transcriptanalyzer/transcript/User_Requests_Intents/Service/TranscriptService.java [hunk 7]
```diff
@@ -112,48 +130,101 @@ private static void flowIterator2(ArrayList<ArrayList<ArrayList<String>>> finalP
...
+
         }
-//        System.out.println(turnsByKey);
+
     }
 
+// End of helper functions to getJsonContent()
 
 //Stores the APIKey in the apiAccess.properties files
 //    //Switched to static
```
Comment (CharlieMorocz) [inline transcript/src/main/java/com/transcriptanalyzer/transcript/User_Requests_Intents/Service/TranscriptService.java:230]: We could probably remove the commented-out code below if it's no longer in use.

**Review by CharlieMorocz — APPROVED** [review]
Overall fantastic job everyone! There was clearly a lot of clever thinking behind how to arrange and analyze this data, and it definitely paid off. The sorting algorithm for the best leaf intents was super impressive. The only concern I had is that some methods were a smidge long (likely by virtue of the complexity of the data we're analyzing) - in the future it may be worth looking into creating some helper methods to shrink them down and improve readability. Mainly though, awesome that this is all working now! Thank you.
